soroban-env-host/src/e2e_invoke.rs

```diff
@@ -533,6 +534,19 @@ pub fn invoke_host_function_in_recording_mode(
     let invoke_result = host.invoke_function(host_function);
     let mut contract_events_and_return_value_size = 0_u32;
     if let Ok(res) = &invoke_result {
+        // In "real" enforcing mode, the module cache gets populated before
+        // invocation and then every access is a hit. In recording mode, we
+        // don't know what contracts will be accessed -- to populate the cache
+        // with -- until after invocation, so we still treat every access as a
+        // hit while the invocation is running and then we charge the population
+        // here _after_ the invocation. It's a bit of a hack but produces the
+        // right cost (on success), just in the wrong order. The main
+        // disadvantage is that invocations that return errors will have
+        // observably-different costs than their real invocations would have;
+        // but we assume people don't care a lot about the cost accuracy of
+        // invocations that return errors.
+        host.maybe_add_module_cache()?;
```

Comment:
Can we do that on exit from the top-level frame, similar to what we do `maybe_emulate_authentication`. This will make the metering a bit more accurate for tests as well. This also is resilient to failures and probably will have a bit better error reporting in case if we run out of budget during populating the cache.

Reply:
Oh, yeah, I think that works. Ok doing so.